tests/simulation/test_particle_tracker.py

```diff
@@ -438,3 +440,141 @@ def test_particle_tracker_stop_particles(request):
         simulation._remove_particles([True, True])
 
     assert np.isnan(simulation.x[0, :]).all()
+
+
+def t_opt(𝜏, t, vd, γd, ν):
+    """
+    fsolve optimization function
+    Eq. 72 in the Friedman paper
+
+    """
+    return t - γd.value**2 / ν.value * (
+        ν.value * 𝜏 - vd.value**2 / const.c.si.value**2 * np.sin(ν.value * 𝜏)
+    )
+
+
+def ExB_trajectory(t, E, B, q=const.e.si, m=const.m_p.si, nonrel=False):
+    """
+    Calculates the relativistically-correct ExB drift trajectory for a
+    particle starting at x=v=0 at t=0 with E in the x direction and
+    B in the y direction and ExB in the z direction. The motion is 2D in the
+    xz plane.
+
+    From https://journals.aps.org/pre/abstract/10.1103/PhysRevE.72.026603
+
+    """
+
+    # TODO: implement this for more regimes?
+    if E >= const.c.si * B:
+        raise ValueError("Currently this function only works for E<cB")
+
+    # Just after Eq. 41
+    vd = (E / B).to(u.m / u.s)
+
+    # Eq. 34
+    ν = q * np.sqrt((const.c.si * B) ** 2 - E**2) / (m * const.c.si)
+
+    if nonrel:
+        # Calculate the positions using the non-relativistic expression
+        # Eq. 79
+        a = m * vd / (q * B)
+        x = (a * (ν * t - np.sin(ν.si.value * t.si.value))).to(u.m)
+        z = (a * (np.cos(ν.si.value * t.si.value) - 1)).to(u.m)
+    else:
+        # Eq. 45
+        γd = 1 / np.sqrt(1 - vd**2 / const.c.si**2)
+
+        # Numerically invert Eq. 72 to calculate the proper time for each time
+        # t in the lab frame
+        𝜏 = np.zeros(t.size)
+        for i, val in enumerate(t):
+            𝜏[i] = fsolve(t_opt, [val.si.value], args=(val.si.value, vd, γd, ν))
+        𝜏 *= u.s
+
+        # Calculate the positions
+        # Eq. 71
+        a = γd * vd / ν
+        x = (a * γd * (ν * 𝜏 - np.sin(ν.si.value * 𝜏.si.value))).to(u.m)
+        z = (a * (np.cos(ν.si.value * 𝜏.si.value) - 1)).to(u.m)
+
+    return x, z
+
+
+def construct_field(
```

Comment:
Personal preference - if this is only used in the below test function, maybe define it as local inside that function?

Reply:
What are your thoughts on moving all of the relevant functions into a class? https://github.com/PlasmaPy/PlasmaPy/blob/2f25b6c847caa2d7a69d815af522f116d8c8772b/tests/simulation/test_particle_tracker.py#L445-L662

Reply:
Looks like a good idea to me!